### PR TITLE
Cinder volume attacher: use instanceID instead of NodeID when verifying attachment

### LINKED ISSUE #39978: Cinder volumes attacher uses NodeName instead of instanceID
**Is this a request for help?**: No

**What keywords did you search in Kubernetes issues before filing this one?**: openstack cinder volume instanceid nodename

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Kubernetes version** (use `kubectl version`): v1.5.1

**Environment**:
- **Cloud provider or hardware configuration**: Openstack Liberty
- **OS** (e.g. from /etc/os-release): Red Hat Enterprise Linux Server 7.2 (Maipo)
- **Kernel** (e.g. `uname -a`): 3.10.0-327.36.2.el7.x86_64 #1 SMP

**What happened**: The Cinder volume was not remounted to other node after POD was rescheduled.

**What you expected to happen**: Cinder volume is detached from previous node mounted on a new one.

**How to reproduce it** (as minimally and precisely as possible):

1. Create a PVC using Cinder;
2. Create a POD under replication controller with one replica using that PVC;
3. Wait for POD to be scheduled and started;
4. Cordon the node where POD is running;
5. Delete the POD;
6. POD is rescheduled to another node, but volume is not attached - it keeps attaching and detaching it to the new node;

**Anything else do we need to know**:

Logs showing that attachment succeeds, but when attacher performs rechecking during reconciliation cycle, it incorrectly decides that volume is not attached any longer:

```
I0113 23:45:20.854316       7 reconciler.go:202] Started AttachVolume for volume "kubernetes.io/cinder/7c0ef8d0-fa65-41b5-b634-098305579935" to node "ak8s-worker-1"
I0113 23:45:22.116749       7 openstack_volumes.go:126] 7c0ef8d0-fa65-41b5-b634-098305579935 kubernetes-dynamic-pvc-6635944c-d431-11e6-a6be-fa163e24f3a8 [map[server_id:b6e0e07c-d64b-45bc-b09d-f8784fa4fbf7 attachment_id:fc236602-3f75-4a6e-beb7-12eeed453cca host_name:<nil> volume_id:7c0ef8d0-fa65-41b5-b634-098305579935 device:/dev/vde id:7c0ef8d0-fa65-41b5-b634-098305579935]]
I0113 23:45:22.116793       7 attacher.go:92] Attach operation is successful. volume "7c0ef8d0-fa65-41b5-b634-098305579935" is already attached to instance "b6e0e07c-d64b-45bc-b09d-f8784fa4fbf7".
I0113 23:45:22.603837       7 openstack_volumes.go:126] 7c0ef8d0-fa65-41b5-b634-098305579935 kubernetes-dynamic-pvc-6635944c-d431-11e6-a6be-fa163e24f3a8 [map[attachment_id:fc236602-3f75-4a6e-beb7-12eeed453cca host_name:<nil> volume_id:7c0ef8d0-fa65-41b5-b634-098305579935 device:/dev/vde id:7c0ef8d0-fa65-41b5-b634-098305579935 server_id:b6e0e07c-d64b-45bc-b09d-f8784fa4fbf7]]
I0113 23:45:22.603879       7 operation_generator.go:217] AttachVolume.Attach succeeded for volume "kubernetes.io/cinder/7c0ef8d0-fa65-41b5-b634-098305579935" (spec.Name: "pvc-6635944c-d431-11e6-a6be-fa163e24f3a8") from node "ak8s-worker-1".
I0113 23:45:22.603967       7 actual_state_of_world.go:447] Report volume "kubernetes.io/cinder/7c0ef8d0-fa65-41b5-b634-098305579935" as attached to node "ak8s-worker-1"
I0113 23:45:22.682490       7 node_status_updater.go:136] Updating status for node "ak8s-worker-1" succeeded. patchBytes: "{}" VolumesAttached: [{kubernetes.io/cinder/7c0ef8d0-fa65-41b5-b634-098305579935 /dev/vde}]
I0113 23:45:23.078355       7 actual_state_of_world.go:343] SetVolumeMountedByNode volume kubernetes.io/cinder/7c0ef8d0-fa65-41b5-b634-098305579935 to the node "ak8s-worker-1" mounted true
I0113 23:45:23.767648       7 openstack_volumes.go:126] 7c0ef8d0-fa65-41b5-b634-098305579935 kubernetes-dynamic-pvc-6635944c-d431-11e6-a6be-fa163e24f3a8 [map[host_name:<nil> volume_id:7c0ef8d0-fa65-41b5-b634-098305579935 device:/dev/vde id:7c0ef8d0-fa65-41b5-b634-098305579935 server_id:b6e0e07c-d64b-45bc-b09d-f8784fa4fbf7 attachment_id:fc236602-3f75-4a6e-beb7-12eeed453cca]]
I0113 23:45:23.767768       7 attacher.go:140] VolumesAreAttached: check volume "7c0ef8d0-fa65-41b5-b634-098305579935" (specName: "pvc-6635944c-d431-11e6-a6be-fa163e24f3a8") is no longer attached
I0113 23:45:23.768069       7 operation_generator.go:162] VerifyVolumesAreAttached determined volume "kubernetes.io/cinder/7c0ef8d0-fa65-41b5-b634-098305579935" (spec.Name: "pvc-6635944c-d431-11e6-a6be-fa163e24f3a8") is no longer attached to node "ak8s-worker-1", therefore it was marked as detached.
I0113 23:45:23.787455       7 reconciler.go:202] Started AttachVolume for volume "kubernetes.io/cinder/7c0ef8d0-fa65-41b5-b634-098305579935" to node "ak8s-worker-1"

```

Later analysis showed that Cinder attacher uses [NodeName, instead of instanceID](https://github.com/kubernetes/kubernetes/blob/6a4d5cd7cc58e28c20ca133dab7b0e9e56192fe3/pkg/volume/cinder/attacher.go#L127), compare e.g. with [Attach operation](https://github.com/kubernetes/kubernetes/blob/6a4d5cd7cc58e28c20ca133dab7b0e9e56192fe3/pkg/volume/cinder/attacher.go#L75-L82) and [DisksAreAttached]( https://github.com/kubernetes/kubernetes/blob/6a4d5cd7cc58e28c20ca133dab7b0e9e56192fe3/pkg/cloudprovider/providers/openstack/openstack_volumes.go#L258)

### PR DESCRIPTION
**Pull request by DukeXar**
**What this PR does / why we need it**: Cinder volume attacher incorrectly uses NodeID instead of openstack instance id, so that reconciliation fails.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #39978 

**Special notes for your reviewer**:

**Release note**:

```release-note
Fixed checking of InstanceID when attaching OpenStack Cinder volume.
```

### CODE REVIEW

**Comment by DukeXar** [comment]
@k8s-ci-robot I signed it!

**Comment by resouer** [comment]
@k8s-bot ok to test

File: pkg/volume/cinder/attacher.go [hunk 1]
```diff
@@ -124,7 +118,13 @@ func (attacher *cinderDiskAttacher) VolumesAreAttached(specs []*volume.Spec, nod
...
 		volumeSpecMap[volumeSource.VolumeID] = spec
 	}
-	attachedResult, err := attacher.cinderProvider.DisksAreAttached(volumeIDList, string(nodeName))
+
+	instanceid, err := attacher.nodeInstanceID(nodeName)
+	if err != nil {
+		return volumesAttachedCheck, err
+	}
+
+	attachedResult, err := attacher.cinderProvider.DisksAreAttached(volumeIDList, instanceid)
```
Comment (resouer) [inline pkg/volume/cinder/attacher.go]: It would be better to have a unit test to verify this chang.

**Comment by DukeXar** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/volume/cinder/attacher.go, line 127 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39998#-KafcTkSvlIVoudiWiOO-r1-127:-KajHc7C4n8GbAVST06h:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/d1ed3630beffa314d408c48ac6692eff700b0aa4/pkg/volume/cinder/attacher.go#L127)):*
<details><summary><i>Previously, resouer (Harry Zhang) wrote…</i></summary><blockquote>

It would be better to have a unit test to verify this chang.
</blockquote></details>

Done.

---

*[pkg/volume/cinder/attacher_test.go, line 194 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39998#-KajHnphsZNhmSL4wtOw:-KajHnphsZNhmSL4wtOx:bouapyt) ([raw file](https://github.com/kubernetes/kubernetes/blob/084d801e0a4d6cdf6eec20941cba7b26b65325b1/pkg/volume/cinder/attacher_test.go#L194)):*
> serializeAttachments

Does not look like the best thing to do, open for suggestions on what is idiomatic in Go

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39998)*

File: pkg/volume/cinder/attacher_test.go [hunk 2]
```diff
@@ -175,6 +181,46 @@ func TestAttachDetach(t *testing.T) {
...
 
+		// Successful VolumesAreAttached call, attached
+		{
+			name:             "VolumesAreAttached_Positive",
+			instanceID:       instanceID,
+			disksAreAttached: disksAreAttachedCall{[]string{diskName}, instanceID, map[string]bool{diskName: true}, nil},
+			test: func(testcase *testcase) (string, error) {
+				attacher := newAttacher(testcase)
+				attachments, err := attacher.VolumesAreAttached([]*volume.Spec{spec}, nodeName)
+				return serializeAttachments(attachments), err
```
Comment (DukeXar) [inline pkg/volume/cinder/attacher_test.go]: > serializeAttachments

Does not look like the best thing to do, open for suggestions on what is idiomatic in Go

File: pkg/volume/cinder/attacher.go (hunk 1 above)
Comment (DukeXar) [inline pkg/volume/cinder/attacher.go]: Done

**Comment by resouer** [comment]
Leave rest of this to @saad-ali for final review, thanks a lot!

**Comment by xsgordon** [comment]
@idvoretskyi can you add the sig/openstack label?

**Comment by jsafrane** [comment]
no response from @saad-ali, I reviewed it by myself
/lgtm
/approve

**Comment by jsafrane** [comment]
@k8s-bot node e2e test this

**Comment by resouer** [comment]
@DukeXar  Congratulations! ㊗️ ! Enjoy the contribution.

**Comment by DukeXar** [comment]
@resouer Thanks. Do I need to cherry-pick it manually to get into the release 1.5?

**Comment by DukeXar** [comment]
@jsafrane does it need to have a milestone and a release note in order to be cherry-picked into the 1.5? I am a bit confused with the process.

**Comment by jsafrane** [comment]
@DukeXar, I added milestone and cherrypick-candidate label, IMO that should be enough

**Pull request merged**